python/ray/serve/master.py

```diff
@@ -1,334 +1,528 @@
+                detached=True,
+                name=SERVE_METRIC_MONITOR_NAME).remote(gc_window_s)
+            # TODO(edoakes): move these into the constructor.
+            start_metric_monitor_loop.remote(self.metric_monitor)
+            self.metric_monitor.add_target.remote(self.router)
 
     def get_metric_monitor(self):
-        assert self.metric_monitor is not None, (
-            "Metric monitor not started yet.")
+        """Returns a handle to the metric monitor managed by this actor."""
         return [self.metric_monitor]
 
+    def _checkpoint(self):
+        """Checkpoint internal state and write it to the KV store."""
+        logger.debug("Writing checkpoint")
+        start = time.time()
+        checkpoint = pickle.dumps(
+            (self.routes, self.backends, self.traffic_policies, self.replicas,
+             self.replicas_to_start, self.replicas_to_stop))
+
+        self.kv_store_client.put("checkpoint", checkpoint)
+        logger.debug("Wrote checkpoint in {:.2f}".format(time.time() - start))
+
+    async def _recover_from_checkpoint(self, checkpoint_bytes):
+        """Recover the cluster state from the provided checkpoint.
+
+        Performs the following operations:
+            1) Deserializes the internal state from the checkpoint.
+            2) Pushes the latest configuration to the HTTP proxy and router
+               in case we crashed before updating them.
+            3) Starts/stops any worker replicas that are pending creation or
+               deletion.
+
+        NOTE: this requires that self.write_lock is already acquired and will
+        release it before returning.
+        """
+        start = time.time()
+        logger.info("Recovering from checkpoint")
+
+        # Load internal state from the checkpoint data.
+        (self.routes, self.backends, self.traffic_policies, self.replicas,
+         self.replicas_to_start,
+         self.replicas_to_stop) = pickle.loads(checkpoint_bytes)
+
+        # Fetch actor handles for all of the backend replicas in the system.
+        # All of these workers are guaranteed to already exist because they
+        # would not be written to a checkpoint in self.workers until they
+        # were created.
+        for backend_tag, replica_tags in self.replicas.items():
+            for replica_tag in replica_tags:
+                self.workers[backend_tag][replica_tag] = ray.util.get_actor(
+                    replica_tag)
+
+        # Push configuration state to the router.
+        # TODO(edoakes): should we make this a pull-only model for simplicity?
+        for endpoint, traffic_policy in self.traffic_policies.items():
+            await self.router.set_traffic.remote(endpoint, traffic_policy)
+
+        for backend_tag, replica_dict in self.workers.items():
+            for replica_tag, worker in replica_dict.items():
+                await self.router.add_new_worker.remote(
+                    backend_tag, replica_tag, worker)
+
```

Comment:
what happen if the router fail somewhere between these lines? router will pull and these will be noop?

Reply:
The router will be restarted and these will retry until they succeed. If they succeed and then router crashes, router will fetch them on startup.